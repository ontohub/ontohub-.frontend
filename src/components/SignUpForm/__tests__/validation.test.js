import {
  addError,
  formHasError,
  validate,
  isUsernameAvailable
} from "../validation";

jest.mock("config/apollo", () => ({
  Client: {
    query: ({ variables: { id } }) => {
      if (id === "bob") {
        return Promise.reject();
      } else {
        return Promise.resolve({ data: { organizationalUnit: id === "ada" } });
      }
    }
  }
}));

describe("addError", () => {
  it("adds an error if the field has no error", () => {
    let errors = { anotherField: ["Some Error message"] };
    addError(true, errors, "someField", "This is the error message");
    expect(errors).toMatchSnapshot();
  });

  it("adds an error if the field already has an error", () => {
    let errors = { someField: ["This is another error message"] };
    addError(true, errors, "someField", "This is the error message");
    expect(errors).toMatchSnapshot();
  });
});

describe("formHasError", () => {
  it("returns true if one of the relevant fields has an error", () => {
    let errors = { name: ["This is an error"] };
    expect(formHasError(errors)).toBeTruthy();
  });

  it("returns false if none of the relevant fields has an error", () => {
    let errors = { otherField: ["This is an error"] };
    expect(formHasError(errors)).toBeFalsy();
  });
});

describe("isUsernameAvailable", () => {
  it("resolves", () => {
    return isUsernameAvailable("ada").then(available => {
      expect(available).toBeFalsy();
    });
  });
});

describe("validate", () => {
  it("on empty fields", () => {
    return validate({})
      .catch(e => e)
      .then(e => {
        expect(e).toMatchSnapshot();
      });
  });

  it("on some non-empty fields", () => {
    return validate({ name: "_._", confirmPassword: "notMatchingPassword" })
      .catch(e => e)
      .then(e => {
        expect(e).toMatchSnapshot();
      });
  });

  it("checks for username availability", () => {
    return validate({
      name: "ada",
      email: "ada@example.com",
      password: "matchingPassword",
      confirmPassword: "matchingPassword"
    })
      .catch(e => e)
      .then(e => {
        expect(e).toMatchSnapshot();
      });
  });

  it("returns an error if availability cannot be checked", () => {
    return validate({
      name: "bob",
      email: "ada@example.com",
      password: "matchingPassword",
      confirmPassword: "matchingPassword"
    })
      .catch(e => e)
      .then(e => {
        expect(e).toMatchSnapshot();
      });
  });
});
