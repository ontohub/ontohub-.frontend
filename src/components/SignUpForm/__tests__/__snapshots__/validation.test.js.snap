// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addError adds an error if the field already has an error 1`] = `
Object {
  "someField": Array [
    "This is another error message",
    "This is the error message",
  ],
}
`;

exports[`addError adds an error if the field has no error 1`] = `
Object {
  "anotherField": Array [
    "Some Error message",
  ],
  "someField": Array [
    "This is the error message",
  ],
}
`;

exports[`setServerErrors calls setErrors with the found errors 1`] = `
Array [
  Array [
    Object {
      "password": Array [
        "Password is missing",
      ],
      "username": Array [
        "Username not found",
      ],
    },
  ],
]
`;

exports[`setServerErrors with an invalid error message calls setErrors with the found errors 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;

exports[`validate checks for username availability 1`] = `
Object {
  "name": Array [
    "Username is already taken",
  ],
  "passwordScore": 10,
}
`;

exports[`validate on empty fields 1`] = `
Object {
  "email": Array [
    "Email is invalid",
  ],
  "name": Array [
    "Username must be at least 3 characters long",
  ],
  "password": Array [
    "Password must be at least 10 characters long",
  ],
  "passwordScore": 10,
}
`;

exports[`validate on some non-empty fields 1`] = `
Object {
  "confirmPassword": Array [
    "Passwords must match",
  ],
  "email": Array [
    "Email is invalid",
  ],
  "name": Array [
    "Username must start and end with a lower case letter or number, and only contain lower case letters, numbers, \\"-\\" and \\"_\\"",
  ],
  "password": Array [
    "Password must be at least 10 characters long",
  ],
  "passwordScore": 10,
}
`;

exports[`validate returns an error if availability cannot be checked 1`] = `
Object {
  "name": Array [
    "Unable to check username availability. Please check your network connection.",
  ],
  "passwordScore": 10,
}
`;
